src/par_iter/chain.rs

```diff
@@ -199,14 +206,58 @@ impl<A, B> Producer for ChainProducer<A, B>
     }
 }
 
-impl<A, B> IntoIterator for ChainProducer<A, B>
-    where A: Producer,
-          B: Producer<Item = A::Item>
+/// ////////////////////////////////////////////////////////////////////////
+/// Wrapper for Chain to implement ExactSizeIterator
+
+pub struct Chain<A, B> {
+    chain: iter::Chain<A, B>,
+    len: usize,
+}
+
+impl<A, B> Chain<A, B> {
+    fn new(a: A, b: B) -> Chain<A, B>
+        where A: ExactSizeIterator,
+              B: ExactSizeIterator<Item=A::Item>
+    {
+        let len = a.len() + b.len();
```

Comment:
AFAICS you didn't write a `size_hint` at all, which the docs say you must for `ExactSizeIterator`.  Then the default `ExactSizeIterator::len()` asserts that the lower and upper bounds are actually equal.  You can override `len` for performance of you know it's always accurate, but in this case I'd just defer to `Chain::size_hint` and let it assert.